File: src/extension/src/scoring/phishing-scorer.ts
import credentialRequestPhrasesData from '../../config/credential-request-phrases.json';
import { brandDomains, shortenerDomains, urgencyKeywords } from '../masters';
import {
  extractDomainFromText,
  extractDomainFromUrl,
  isIpAddressDomain,
  isPunycodeDomain,
  levenshteinDistance,
} from './utils';
import type { ScoreReason, ScoreResult, ScoringEmailInput } from './types';

interface CredentialRequestPhrase {
  id: number;
  phrase: string;
  lang: 'ja' | 'en';
}

const credentialRequestPhrases: CredentialRequestPhrase[] =
  credentialRequestPhrasesData as CredentialRequestPhrase[];

const URGENCY_KEYWORD_POINTS_PER_MATCH = 5;
const URGENCY_KEYWORD_POINTS_MAX = 20;
const SCORE_MIN = 0;
const SCORE_MAX = 100;
const BRAND_LOOKALIKE_MIN_DISTANCE = 1;
const BRAND_LOOKALIKE_MAX_DISTANCE = 2;

function stripSubdomain(domain: string): string {
  return domain.toLowerCase();
}

function isSameOrSubdomain(domain: string, officialDomain: string): boolean {
  const normalizedDomain = stripSubdomain(domain);
  const normalizedOfficial = stripSubdomain(officialDomain);
  return normalizedDomain === normalizedOfficial || normalizedDomain.endsWith(`.${normalizedOfficial}`);
}

export class PhishingScorer {
  constructor(private readonly email: ScoringEmailInput) {}

  evalAuthHeaders(): ScoreReason[] {
    const headers = this.email.authHeaders;
    if (!headers) {
      return [{ code: 'AUTH_HEADERS_UNAVAILABLE', delta: 0 }];
    }

    const values = [headers.spf, headers.dkim, headers.dmarc];
    if (values.every((value) => value === 'pass')) {
      return [{ code: 'AUTH_HEADERS_ALL_PASS', delta: -30 }];
    }
    if (values.some((value) => value === 'fail')) {
      return [{ code: 'AUTH_HEADERS_ANY_FAIL', delta: 30 }];
    }
    return [{ code: 'AUTH_HEADERS_MIXED', delta: 0 }];
  }

  evalUrls(): ScoreReason[] {
    const reasons: ScoreReason[] = [];
    const links = this.email.links;

    const hasMismatch = links.some((link) => {
      const displayDomain = extractDomainFromText(link.displayText);
      const hrefDomain = extractDomainFromUrl(link.href);
      return displayDomain !== null && hrefDomain !== null && displayDomain !== hrefDomain;
    });
    if (hasMismatch) {
      reasons.push({ code: 'URL_DISPLAY_HREF_MISMATCH', delta: 25 });
    }

    const hasShortener = links.some((link) => {
      const hrefDomain = extractDomainFromUrl(link.href);
      return hrefDomain !== null && shortenerDomains.some((s) => s.domain === hrefDomain);
    });
    if (hasShortener) {
      reasons.push({ code: 'URL_SHORTENER_DOMAIN', delta: 10 });
    }

    const hasPunycodeOrIp = links.some((link) => {
      const hrefDomain = extractDomainFromUrl(link.href);
      return hrefDomain !== null && (isPunycodeDomain(hrefDomain) || isIpAddressDomain(hrefDomain));
    });
    if (hasPunycodeOrIp) {
      reasons.push({ code: 'URL_PUNYCODE_OR_IP', delta: 25 });
    }

    const hasBrandLookalike = links.some((link) => {
      const hrefDomain = extractDomainFromUrl(link.href);
      if (hrefDomain === null) {
        return false;
      }
      return brandDomains.some((brand) => {
        if (hrefDomain === brand.official_domain) {
          return false;
        }
        const distance = levenshteinDistance(hrefDomain, brand.official_domain);
        return distance >= BRAND_LOOKALIKE_MIN_DISTANCE && distance <= BRAND_LOOKALIKE_MAX_DISTANCE;
      });
    });
    if (hasBrandLookalike) {
      reasons.push({ code: 'URL_BRAND_LOOKALIKE_DOMAIN', delta: 30 });
    }

    return reasons;
  }

  evalUrgency(): ScoreReason[] {
    const text = `${this.email.subject} ${this.email.body}`.toLowerCase();
    const matchCount = urgencyKeywords.filter((keyword) => text.includes(keyword.keyword.toLowerCase()))
      .length;
    if (matchCount === 0) {
      return [];
    }
    const delta = Math.min(matchCount * URGENCY_KEYWORD_POINTS_PER_MATCH, URGENCY_KEYWORD_POINTS_MAX);
    return [{ code: 'URGENCY_KEYWORDS_MATCHED', delta }];
  }

  evalCredentialRequest(): ScoreReason[] {
    const text = `${this.email.subject} ${this.email.body}`.toLowerCase();
    const matched = credentialRequestPhrases.some((phrase) => text.includes(phrase.phrase.toLowerCase()));
    return matched ? [{ code: 'CREDENTIAL_REQUEST_DETECTED', delta: 20 }] : [];
  }

  evalBrandMismatch(): ScoreReason[] {
    const displayName = this.email.senderDisplayName.toLowerCase();
    const matchedBrand = brandDomains.find((brand) => displayName.includes(brand.brand_name.toLowerCase()));
    if (!matchedBrand) {
      return [];
    }
    if (isSameOrSubdomain(this.email.senderDomain, matchedBrand.official_domain)) {
      return [];
    }
    return [{ code: 'SENDER_BRAND_DOMAIN_MISMATCH', delta: 25 }];
  }

  score(): ScoreResult {
    const reasons = [
      ...this.evalAuthHeaders(),
      ...this.evalUrls(),
      ...this.evalUrgency(),
      ...this.evalCredentialRequest(),
      ...this.evalBrandMismatch(),
    ];
    const rawScore = reasons.reduce((total, reason) => total + reason.delta, 0);
    const score = Math.min(SCORE_MAX, Math.max(SCORE_MIN, rawScore));
    return { score, reasons };
  }
}

File: src/api/app/services/rule_based/phishing_scorer.rb
require "json"

# src/extension/src/scoring/phishing-scorer.ts のRuby移植版。
# requirements.md 1.4-A準拠。サーバ側でクライアント計算のスコアを信頼せず、
# 抽出データから独自に再計算する(OWASP A04対策)。
module RuleBased
  class PhishingScorer
    Reason = Struct.new(:code, :delta) do
      def to_h
        { code: code, delta: delta }
      end
    end

    CREDENTIAL_REQUEST_PHRASES_PATH = Rails.root.join(
      "..", "extension", "config", "credential-request-phrases.json",
    )
    CREDENTIAL_REQUEST_PHRASES = JSON.parse(File.read(CREDENTIAL_REQUEST_PHRASES_PATH)).freeze

    URGENCY_KEYWORD_POINTS_PER_MATCH = 5
    URGENCY_KEYWORD_POINTS_MAX = 20
    SCORE_MIN = 0
    SCORE_MAX = 100
    BRAND_LOOKALIKE_MIN_DISTANCE = 1
    BRAND_LOOKALIKE_MAX_DISTANCE = 2

    def initialize(email)
      @email = email
    end

    def eval_auth_headers
      headers = @email[:auth_headers]
      return [ Reason.new("AUTH_HEADERS_UNAVAILABLE", 0) ] if headers.nil?

      values = [ headers[:spf], headers[:dkim], headers[:dmarc] ]
      if values.all? { |v| v == "pass" }
        [ Reason.new("AUTH_HEADERS_ALL_PASS", -30) ]
      elsif values.any? { |v| v == "fail" }
        [ Reason.new("AUTH_HEADERS_ANY_FAIL", 30) ]
      else
        [ Reason.new("AUTH_HEADERS_MIXED", 0) ]
      end
    end

    def eval_urls
      reasons = []
      links = @email[:links] || []

      if links.any? { |link| mismatched_display_href?(link) }
        reasons << Reason.new("URL_DISPLAY_HREF_MISMATCH", 25)
      end

      if links.any? { |link| shortener_domain?(link) }
        reasons << Reason.new("URL_SHORTENER_DOMAIN", 10)
      end

      if links.any? { |link| punycode_or_ip?(link) }
        reasons << Reason.new("URL_PUNYCODE_OR_IP", 25)
      end

      if links.any? { |link| brand_lookalike?(link) }
        reasons << Reason.new("URL_BRAND_LOOKALIKE_DOMAIN", 30)
      end

      reasons
    end

    def eval_urgency
      text = "#{@email[:subject]} #{@email[:body]}".downcase
      match_count = urgency_keywords.count { |keyword| text.include?(keyword.downcase) }
      return [] if match_count.zero?

      delta = [ match_count * URGENCY_KEYWORD_POINTS_PER_MATCH, URGENCY_KEYWORD_POINTS_MAX ].min
      [ Reason.new("URGENCY_KEYWORDS_MATCHED", delta) ]
    end

    def eval_credential_request
      text = "#{@email[:subject]} #{@email[:body]}".downcase
      matched = CREDENTIAL_REQUEST_PHRASES.any? { |p| text.include?(p["phrase"].downcase) }
      matched ? [ Reason.new("CREDENTIAL_REQUEST_DETECTED", 20) ] : []
    end

    def eval_brand_mismatch
      display_name = (@email[:sender_display_name] || "").downcase
      matched_brand = brand_domains.find { |brand| display_name.include?(brand.brand_name.downcase) }
      return [] if matched_brand.nil?
      return [] if same_or_subdomain?(@email[:sender_domain].to_s, matched_brand.official_domain)

      [ Reason.new("SENDER_BRAND_DOMAIN_MISMATCH", 25) ]
    end

    def score
      reasons = eval_auth_headers + eval_urls + eval_urgency + eval_credential_request + eval_brand_mismatch
      raw_score = reasons.sum(&:delta)
      { score: raw_score.clamp(SCORE_MIN, SCORE_MAX), reasons: reasons.map(&:to_h) }
    end

    private

    def mismatched_display_href?(link)
      display_domain = DomainUtils.extract_domain_from_text(link[:display_text].to_s)
      href_domain = DomainUtils.extract_domain_from_url(link[:href].to_s)
      !display_domain.nil? && !href_domain.nil? && display_domain != href_domain
    end

    def shortener_domain?(link)
      href_domain = DomainUtils.extract_domain_from_url(link[:href].to_s)
      !href_domain.nil? && shortener_domains.include?(href_domain)
    end

    def punycode_or_ip?(link)
      href_domain = DomainUtils.extract_domain_from_url(link[:href].to_s)
      return false if href_domain.nil?

      DomainUtils.punycode_domain?(href_domain) || DomainUtils.ip_address_domain?(href_domain)
    end

    def brand_lookalike?(link)
      href_domain = DomainUtils.extract_domain_from_url(link[:href].to_s)
      return false if href_domain.nil?

      brand_domains.any? do |brand|
        next false if href_domain == brand.official_domain

        distance = DomainUtils.levenshtein_distance(href_domain, brand.official_domain)
        distance >= BRAND_LOOKALIKE_MIN_DISTANCE && distance <= BRAND_LOOKALIKE_MAX_DISTANCE
      end
    end

    def same_or_subdomain?(domain, official_domain)
      normalized_domain = domain.downcase
      normalized_official = official_domain.downcase
      normalized_domain == normalized_official || normalized_domain.end_with?(".#{normalized_official}")
    end

    def urgency_keywords
      @urgency_keywords ||= UrgencyKeyword.pluck(:keyword)
    end

    def brand_domains
      @brand_domains ||= BrandDomain.all.to_a
    end

    def shortener_domains
      @shortener_domains ||= ShortenerDomain.pluck(:domain)
    end
  end
end
